File: apps/backend/app/schemas/chatbot.py
"""Schemas for chatbot API requests and responses."""
import uuid
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field


class ChatMessagePayload(BaseModel):
    """Client-provided chat message."""

    role: str = Field(..., examples=["system", "user", "assistant"])
    content: str = Field(..., min_length=0)


class ContextTracePayload(BaseModel):
    """Context data that informed the assistant response."""

    rag_context: list[str] = Field(default_factory=list)
    image_contexts: list[str] = Field(default_factory=list)
    conversation_history: list[str] = Field(default_factory=list)
    merged_context: str | None = None


class TeachingImagePayload(BaseModel):
    """Client-provided teaching screenshot to pass to multimodal LLM."""

    mime_type: Literal["image/png", "image/jpeg", "image/webp", "image/gif"]
    data_base64: str = Field(..., min_length=1)
    description: str | None = Field(default=None, max_length=500)
    source: str | None = Field(default="screen_capture", max_length=100)


class ConversationSummaryPayload(BaseModel):
    """Conversation metadata for listing conversations per learner."""

    id: uuid.UUID
    title: str | None = None
    course_id: uuid.UUID | None = None
    lesson_id: uuid.UUID | None = None
    is_active: bool
    updated_at: datetime


class ConversationMessagePayload(BaseModel):
    """Persisted message in a conversation thread."""

    id: uuid.UUID
    role: str
    content: str
    model_version: str | None = None
    tokens_used: int | None = None
    created_at: datetime


class ConversationMessagesResponse(BaseModel):
    """Message list for one conversation thread."""

    conversation_id: uuid.UUID
    messages: list[ConversationMessagePayload] = Field(default_factory=list)


class ChatbotAskRequest(BaseModel):
    """Ask request payload for chatbot endpoint."""

    question: str = Field(..., min_length=1)
    conversation_id: uuid.UUID | None = None
    conversation_title: str | None = Field(default=None, max_length=255)
    course_id: uuid.UUID | None = None
    lesson_id: uuid.UUID | None = None
    history: list[ChatMessagePayload] = Field(default_factory=list)
    context_docs: list[str] = Field(default_factory=list)
    image_contexts: list[str] = Field(default_factory=list)
    teaching_images: list[TeachingImagePayload] = Field(default_factory=list)
    system_prompt: str | None = None
    temperature: float | None = Field(default=None, ge=0.0, le=2.0)
    max_output_tokens: int | None = Field(default=None, ge=1, le=8192)
    thinking_level: Literal["low", "medium", "high"] | None = None


class AssignmentChatbotAskRequest(BaseModel):
    """Ask request payload for assignment support chatbot endpoint."""

    question: str = Field(..., min_length=1)
    assignment_id: uuid.UUID
    conversation_id: uuid.UUID | None = None
    history: list[ChatMessagePayload] = Field(default_factory=list)
    temperature: float | None = Field(default=None, ge=0.0, le=2.0)
    max_output_tokens: int | None = Field(default=None, ge=1, le=8192)
    thinking_level: Literal["low", "medium", "high"] | None = None


class AssignmentChatbotPreloadRequest(BaseModel):
    """Request payload for warming assignment chatbot context."""

    assignment_id: uuid.UUID


class AssignmentChatbotPreloadResponse(BaseModel):
    """Response payload for assignment chatbot context preload."""

    assignment_id: uuid.UUID
    question_count: int
    context_length: int


class ChatbotAskResponse(BaseModel):
    """Chatbot response payload."""

    model_config = ConfigDict(from_attributes=True)

    answer: str
    conversation_id: uuid.UUID
    conversation_title: str | None = None
    provider: str
    model: str
    finish_reason: str | None = None
    usage: dict[str, int] | None = None
    messages: list[ChatMessagePayload] = Field(default_factory=list)
    context: ContextTracePayload | None = None

File: apps/frontend/src/lib/api/chatbot.ts
import { apiClient } from "./client";

const CHATBOT_REQUEST_TIMEOUT_MS = 4 * 60 * 1000;

export interface ChatMessagePayload {
  role: string | "system" | "user" | "assistant";
  content: string;
}

export interface TeachingImagePayload {
  mime_type: "image/png" | "image/jpeg" | "image/webp" | "image/gif";
  data_base64: string;
  description?: string;
  source?: string;
}

export interface ChatbotAskRequest {
  question: string;
  conversation_id?: string;
  conversation_title?: string;
  course_id?: string;
  lesson_id?: string;
  history?: ChatMessagePayload[];
  context_docs?: string[];
  image_contexts?: string[];
  teaching_images?: TeachingImagePayload[];
  temperature?: number;
  max_output_tokens?: number;
  thinking_level?: "low" | "medium" | "high";
}

export interface AssignmentChatbotAskRequest {
  question: string;
  assignment_id: string;
  conversation_id?: string;
  history?: ChatMessagePayload[];
  temperature?: number;
  max_output_tokens?: number;
  thinking_level?: "low" | "medium" | "high";
}

export interface AssignmentChatbotPreloadRequest {
  assignment_id: string;
}

export interface AssignmentChatbotPreloadResponse {
  assignment_id: string;
  question_count: number;
  context_length: number;
}

export interface ChatbotAskResponse {
  answer: string;
  conversation_id: string;
  conversation_title?: string;
  provider: string;
  model: string;
  finish_reason?: string;
  usage?: Record<string, number>;
  messages?: ChatMessagePayload[];
}

export interface ConversationSummaryPayload {
  id: string;
  title?: string;
  course_id?: string;
  lesson_id?: string;
  is_active: boolean;
  updated_at: string;
}

export interface ConversationMessagePayload {
  id: string;
  role: string;
  content: string;
  model_version?: string;
  tokens_used?: number;
  created_at: string;
}

export interface ConversationMessagesResponse {
  conversation_id: string;
  messages: ConversationMessagePayload[];
}

export const chatbotApi = {
  ask: async (data: ChatbotAskRequest): Promise<ChatbotAskResponse> => {
    const response = await apiClient.post<ChatbotAskRequest, any>("/chatbot/ask", data, {
      timeout: CHATBOT_REQUEST_TIMEOUT_MS,
    });
    return response.data;
  },

  askAssignment: async (data: AssignmentChatbotAskRequest): Promise<ChatbotAskResponse> => {
    const response = await apiClient.post<AssignmentChatbotAskRequest, any>("/chatbot/assignment/ask", data, {
      timeout: CHATBOT_REQUEST_TIMEOUT_MS,
    });
    return response.data;
  },

  preloadAssignment: async (data: AssignmentChatbotPreloadRequest): Promise<AssignmentChatbotPreloadResponse> => {
    const response = await apiClient.post<AssignmentChatbotPreloadRequest, any>("/chatbot/assignment/preload", data);
    return response.data;
  },

  getConversations: async (limit: number = 20): Promise<ConversationSummaryPayload[]> => {
    const response = await apiClient.get<ConversationSummaryPayload[]>(`/chatbot/conversations?limit=${limit}`);
    return response.data;
  },

  getConversationMessages: async (conversationId: string): Promise<ConversationMessagesResponse> => {
    const response = await apiClient.get<ConversationMessagesResponse>(`/chatbot/conversations/${conversationId}/messages`);
    return response.data;
  },

  getProviders: async (): Promise<any> => {
    const response = await apiClient.get<any>("/chatbot/providers");
    return response.data;
  },
};
